python/lsst/ts/ofc/controllers/zernike_controller.py

```diff
@@ -0,0 +1,356 @@
+
+    @property
+    def aggregated_state(self) -> np.ndarray[float]:
+        """Returns the aggregated state.
+
+        Returns
+        -------
+        `np.ndarray[float]`
+            Aggregated state.
+        """
+        return self.zk_state
+
+    def set_aggregated_state(self, value: np.ndarray[float]) -> None:
+        """Set the aggregated state.
+
+        Parameters
+        ----------
+        value : `np.ndarray[float]`
+            Aggregated state.
+        """
+        self.zk_state += value
+
+    def reset_history(self) -> None:
+        """Reset the history of the controller."""
+        self.zk_state0 = self.zk_state.copy()
+        self.integral = self.setpoint - self.zk_state
+        self.previous_error = self.integral.copy()
+        self.filtered_derivative = np.zeros_like(self.zk_state)
+
+    def control_step(
+        self,
+        filter_name: str,
+        wfe: np.ndarray[float],
+        sensor_names: list[str],
+        rotation_angle: float = 0.0,
+    ) -> np.ndarray[float]:
+        """Estimate uk in the basis of degree of freedom (DOF).
+
+        Parameters
+        ----------
+        filter_name : `string`
+            Name of the filter.
+        zk_state : `numpy.ndarray`
+            Optical state in the basis of DOF.
+        sensor_names : `list` [`string`]
+            List of sensor names.
+        rotation_angle : `float`
+            Rotation angle in degrees.
+
+        Raises
+        ------
+        NotImplementedError
+            Child class should implemented this.
+        """
+        wfe = np.array(wfe)
+        for idx in range(wfe.shape[0]):
+            wfe_sensor = np.pad(wfe[idx, :], (self.ofc_data.znmin, 0))
+
+            zk_galsim = galsim.zernike.Zernike(
+                wfe_sensor,
+                R_outer=self.ofc_data.config["pupil"]["radius_outer"],
+                R_inner=self.ofc_data.config["pupil"]["radius_inner"],
+            )
+            # Note that we need to remove the first 4 Zernike coefficients
+            # since we padded the wfe with zeros for the 4 first Zernike
+            wfe[idx, :] = zk_galsim.rotate(np.deg2rad(rotation_angle)).coef[
+                self.ofc_data.znmin :
+            ]
+
+        self.log.debug(f"Derotated wavefront error: {wfe}")
+        # Compute wavefront error deviation from the intrinsic wavefront error
+        # y = wfe - intrinsic_zk - y2_correction
+        # y2_correction is a static correction for the
+        # deviation currently set to zero.
+        y2_correction = np.array(
+            [self.ofc_data.y2_correction[sensor] for sensor in sensor_names]
+        )
+        y = (
+            wfe
+            - get_intrinsic_zernikes(
+                self.ofc_data, filter_name, sensor_names, rotation_angle
+            )
+            - y2_correction
+        )
+        conversion_factors = self.getPsfGradPerZernike(jmax=28)
+
+        zernike_step = self.calculate_pid_step(y * conversion_factors[np.newaxis, :])
```

Comment:
I think you should apply this to the wfe only no? instead of using the entire aggregated state?